src/Microsoft.IdentityModel.Tokens/AsymmetricAdapter.cs

```diff
@@ -146,64 +103,65 @@ internal AsymmetricAdapter(SecurityKey key, string algorithm, HashAlgorithm hash
             // If we use the RsaParameters, we create a new RSA object and will need to dispose.
             if (key is RsaSecurityKey rsaKey)
             {
-                if (rsaKey.Rsa != null)
-                    Initialize(rsaKey.Rsa, algorithm);
-                else
-                {
-                    var rsa = RSA.Create();
-                    rsa.ImportParameters(rsaKey.Parameters);
-                    Initialize(rsa, algorithm);
-                    _disposeCryptoOperators = true;
-                }
+                IntializeUsingRsaSecurityKey(rsaKey, algorithm);
             }
             else if (key is X509SecurityKey x509Key)
             {
-#if NET45 || NET451
-                if (requirePrivateKey)
-                    Initialize(x509Key.Certificate.PrivateKey as RSA, algorithm);
-                else
-                    Initialize(x509Key.Certificate.PublicKey.Key as RSA, algorithm);
-#endif
-
-#if NET461 || NETSTANDARD1_4 || NETSTANDARD2_0
-                if (requirePrivateKey)
-                    Initialize(x509Key.Certificate.GetRSAPrivateKey(), algorithm);
-                else
-                    Initialize(x509Key.Certificate.GetRSAPublicKey(), algorithm);
-#endif
+                InitializeUsingX509SecurityKey(x509Key, algorithm, requirePrivateKey);
             }
             else if (key is JsonWebKey jsonWebKey)
             {
-                if (jsonWebKey.Kty == JsonWebAlgorithmsKeyTypes.RSA)
+                if (JsonWebKeyConverter.TryConvertToSecurityKey(jsonWebKey, out SecurityKey securityKey))
```

Comment:
If a jsonWebKey can't be converted to a security key, Signature/Verify functions won't be set.
When a user calls signatureProvider.Sign() we would throw NotSupportedException with "Unable to Sign, Internal SignFunction is not available."

I think that we should be more explicit and throw an exception here if we see that we can't covert a jsonWebKey to a security key.

Reply:
I had originally changed to follow that logic, but choose a different exception as it made more sense.
ill try again and be more careful with the exception and see where that goes.